p9/handlers.go

```diff
@@ -970,6 +970,22 @@ func (t *tstatfs) handle(cs *connState) message {
 	return &rstatfs{st}
 }
 
+// handle implements handler.handle.
+func (t *tlock) handle(cs *connState) message {
+	// Lookup the fid.
+	ref, ok := cs.LookupFID(t.fid)
+	if !ok {
+		return newErr(linux.EBADF)
+	}
+	defer ref.DecRef()
+
+	r1 := lockOK
+	if err := ref.file.Lock(int(t.PID), int(t.Type), int(t.Flags), t.Start, t.Length, t.Client); err != nil {
```

Comment:
I like what you're trying to do here. You don't want clients to return (Status, error), which would be a total non-Go-ism. But in turn, (1) you're now not allowing clients to allow arbitrary error strings. And (2), clients are now also not allowed to return LockGrace or LockBlocked. 

How would you feel about making & documenting special error values?

var ErrLockBlocked = errors.New(...)
var ErrLockError = errors.New(...)
var ErrLockGrace = errors.New(...)

Document them in the p9.File.Lock docstring? I don't know if users will read that... it'd be clearer how to return them if they _were_ separate. Which sucks, I'm sorry.

Reply:
When are clients supposed to return Rlock{status: Error} versus Rlerror{"some string"}? That's totally confusing

Reply:
I never really liked 9P2000.L ...

Reply:
I have to reread your idea, my brain just quit for the night :-) WIll try to wrap this tomorrow.

Reply:
OK, thinking about this
There is the question of whether the lock was able to succeed (i.e. stuff like ENOSYS) and the question the result of the lock operation, i.e. blocked, error, grace. 

So it's the old "could you do it or not" and the "if you could do it, how did it work out"

you might fail with a bad fid, or bad offset, or ... 

or everything might have been fine, but the file was locked, or you asked for an async operation and it said "no, sorry, that will block"

In the client side code on linux, the 9p2000.L error and the status return from the op itself go different directions. 
Hence this design. The lock status is supposed to be part of the return, and Rlerror should only contain errors that occurred "could you do it or not" not "how did it go"
So there are two return codes for lock.

Reply:
put another way, this status byte is very bit as much a return value for the lock as a string for readsymlink. readlink can return an error, or no error and an empty string. This is kind of that same thing.

Reply:
in fact here's the linux client code
```
int p9_client_lock_dotl(struct p9_fid *fid, struct p9_flock *flock, u8 *status)
...
        if (IS_ERR(req))
                return PTR_ERR(req);

        err = p9pdu_readf(&req->rc, clnt->proto_version, "b", status);
        if (err) {
                trace_9p_protocol_dump(clnt, &req->rc);
                goto error;
        }
        p9_debug(P9_DEBUG_9P, "<<< RLOCK status %i\n", *status);
```

caller: 
```
        /* map 9p status to VFS status */
        switch (status) {
        case P9_LOCK_SUCCESS:
                res = 0;
                break;
        case P9_LOCK_BLOCKED:
                res = -EAGAIN;
                break;
        default:
                WARN_ONCE(1, "unknown lock status code: %d\n", status);
                /* fall through */
        case P9_LOCK_ERROR:
        case P9_LOCK_GRACE:
                res = -ENOLCK;
                break;
        }
```

I don't see another way to do this, given the way the kernel client handles it.

Reply:
also. by clients above, do you mean servers? 

I agree on the issue of arbitrary strings, but the kernel client really wants errors and status distinguished ...

Reply:
Yeah, you might be right. If you want, we can do either the current way or the way you had before. Do you want to give people the option of returning Blocked? That's currently not possible.

Reply:
servers can return errors, it's this condition: https://elixir.bootlin.com/linux/v6.1/source/fs/9p/vfs_file.c#L181

```
		res = p9_client_lock_dotl(fid, &flock, &status);
		if (res < 0)
			goto out_unlock;
```

then the status gets evaluated later in the switch like you quoted already.

I added the LockStatus back